From 2e8bff4dcede0daa480168de779050535db85b97 Mon Sep 17 00:00:00 2001
From: Bin Hu <hubin73@huawei.com>
Date: Thu, 17 Feb 2022 23:23:55 +0800
Subject: [PATCH 1/3] livepatch: support force enable or disable function

some functions in IO path are called frequently, in this case,
patches can not be enabled due to stack check failing. We use
the klp_func->force to indicate functions that can be
enabled/disabled without stack check. Also, we add a new patch
state forcedisabled to indicate the patch being disabled and
could not being rmmod.

Signed-off-by: Zhipeng Xie <xiezhipeng1@huawei.com>
Signed-off-by: Bin Hu <hubin73@huawei.com>
---
 arch/arm64/kernel/livepatch.c |  42 ++++++++-
 arch/x86/kernel/livepatch.c   |  69 ++++++++++++++-
 include/linux/livepatch.h     |  20 ++++-
 kernel/livepatch/Kconfig      |  12 +++
 kernel/livepatch/core.c       | 158 ++++++++++++++++++++++++++++++++++
 5 files changed, 298 insertions(+), 3 deletions(-)

diff --git a/arch/arm64/kernel/livepatch.c b/arch/arm64/kernel/livepatch.c
index 363fb8e41..8aefac0d7 100644
--- a/arch/arm64/kernel/livepatch.c
+++ b/arch/arm64/kernel/livepatch.c
@@ -73,7 +73,11 @@ static bool klp_check_jump_func(void *ws_args, unsigned long pc)
 {
 	struct walk_stackframe_args *args = ws_args;
 
+#ifdef CONFIG_EULEROS_LIVEPATCH_FORCE_PATCH
+	return args->check_func(args->data, &args->ret, pc, args->enable, &args->unloadable);
+#else
 	return args->check_func(args->data, &args->ret, pc);
+#endif
 }
 
 static int check_task_calltrace(struct task_struct *t,
@@ -81,6 +85,9 @@ static int check_task_calltrace(struct task_struct *t,
 				bool (*fn)(void *, unsigned long))
 {
 	struct stackframe frame;
+#ifdef CONFIG_EULEROS_LIVEPATCH_FORCE_PATCH
+	int last;
+#endif
 
 	/*
 	 * Handle the current carefully on each CPUs, we shouldn't
@@ -98,7 +105,16 @@ static int check_task_calltrace(struct task_struct *t,
 		frame.pc = thread_saved_pc(t);
 	}
 	start_backtrace(&frame, frame.fp, frame.pc);
+#ifdef CONFIG_EULEROS_LIVEPATCH_FORCE_PATCH
+	last = args->unloadable;
 	walk_stackframe(t, &frame, fn, args);
+	if (last != args->unloadable) {
+		pr_info("PID: %d Comm: %.20s\n", t->pid, t->comm);
+		show_stack(t, NULL, KERN_INFO);
+	}
+#else
+	walk_stackframe(t, &frame, fn, args);
+#endif
 	if (args->ret) {
 		pr_info("PID: %d Comm: %.20s\n", t->pid, t->comm);
 		show_stack(t, NULL, KERN_INFO);
@@ -130,9 +146,15 @@ static int do_check_calltrace(struct walk_stackframe_args *args,
 }
 
 #ifdef CONFIG_LIVEPATCH_BREAKPOINT_NO_STOP_MACHINE
+#ifdef CONFIG_EULEROS_LIVEPATCH_FORCE_PATCH
+int arch_klp_check_task_calltrace(struct task_struct *t,
+				  bool (*check_func)(void *, int *, unsigned long, int, int *),
+				  void *data)
+#else
 int arch_klp_check_task_calltrace(struct task_struct *t,
 				  bool (*check_func)(void *, int *, unsigned long),
 				  void *data)
+#endif
 {
 	struct walk_stackframe_args args = {
 		.data = data,
@@ -146,15 +168,33 @@ int arch_klp_check_task_calltrace(struct task_struct *t,
 }
 #endif
 
+#ifdef CONFIG_EULEROS_LIVEPATCH_FORCE_PATCH
+int arch_klp_check_calltrace(bool (*check_func)(void *, int *, unsigned long, int, int *), void *data, int enable, int *unloadable)
+#else
 int arch_klp_check_calltrace(bool (*check_func)(void *, int *, unsigned long), void *data)
+#endif
 {
+#ifdef CONFIG_EULEROS_LIVEPATCH_FORCE_PATCH
+	int ret = 0;
+#endif
 	struct walk_stackframe_args args = {
-		.data = data,
 		.ret = 0,
+		.data = data,
 		.check_func = check_func,
+#ifdef CONFIG_EULEROS_LIVEPATCH_FORCE_PATCH
+		.enable = enable,
+		.unloadable = 1,
+#endif
 	};
 
+#ifdef CONFIG_EULEROS_LIVEPATCH_FORCE_PATCH
+	ret = do_check_calltrace(&args, klp_check_jump_func);
+	if (unloadable)
+		*unloadable = args.unloadable;
+	return ret;
+#else
 	return do_check_calltrace(&args, klp_check_jump_func);
+#endif
 }
 
 static bool check_module_calltrace(void *ws_args, unsigned long pc)
diff --git a/arch/x86/kernel/livepatch.c b/arch/x86/kernel/livepatch.c
index 99b726296..961a8368e 100644
--- a/arch/x86/kernel/livepatch.c
+++ b/arch/x86/kernel/livepatch.c
@@ -103,8 +103,13 @@ static void klp_print_stack_trace(void *trace_ptr, int trace_len)
 #endif
 #define MAX_STACK_ENTRIES  100
 
+#ifdef CONFIG_EULEROS_LIVEPATCH_FORCE_PATCH
+static int klp_check_stack(void *trace_ptr, int trace_len,
+			   bool (*fn)(void *, int *, unsigned long, int, int *), void *data, int enable, int *unloadable)
+#else
 static int klp_check_stack(void *trace_ptr, int trace_len,
 			   bool (*fn)(void *, int *, unsigned long), void *data)
+#endif
 {
 #ifdef CONFIG_ARCH_STACKWALK
 	unsigned long *trace = trace_ptr;
@@ -121,11 +126,22 @@ static int klp_check_stack(void *trace_ptr, int trace_len,
 	for (i = 0; i < trace->nr_entries; i++) {
 		address = trace->entries[i];
 #endif
+#ifdef CONFIG_EULEROS_LIVEPATCH_FORCE_PATCH
+		if (!fn(data, &ret, address, enable, unloadable)) {
+#else
 		if (!fn(data, &ret, address)) {
+#endif
 #ifdef CONFIG_ARCH_STACKWALK
 			klp_print_stack_trace(trace_ptr, trace_len);
 #else
 			klp_print_stack_trace(trace_ptr, 0);
+#endif
+#ifdef CONFIG_EULEROS_LIVEPATCH_FORCE_PATCH
+			if (ret == -EBUSY && unloadable) {
+				*unloadable = 0;
+				ret = 0;
+				continue;
+			}
 #endif
 			return ret;
 		}
@@ -134,9 +150,15 @@ static int klp_check_stack(void *trace_ptr, int trace_len,
 	return 0;
 }
 
+#ifdef CONFIG_EULEROS_LIVEPATCH_FORCE_PATCH
+static int check_task_calltrace(struct task_struct *t,
+				bool (*fn)(void *, int *, unsigned long, int, int *),
+				void *data, int enable, int *unloadable)
+#else
 static int check_task_calltrace(struct task_struct *t,
 				bool (*fn)(void *, int *, unsigned long),
 				void *data)
+#endif
 {
 	int ret = 0;
 	static unsigned long trace_entries[MAX_STACK_ENTRIES];
@@ -154,7 +176,11 @@ static int check_task_calltrace(struct task_struct *t,
 		return ret;
 	}
 	trace_len = ret;
+#ifdef CONFIG_EULEROS_LIVEPATCH_FORCE_PATCH
+	ret = klp_check_stack(trace_entries, trace_len, fn, data, enable, unloadable);
+#else
 	ret = klp_check_stack(trace_entries, trace_len, fn, data);
+#endif
 #else
 	trace.skip = 0;
 	trace.nr_entries = 0;
@@ -166,7 +192,11 @@ static int check_task_calltrace(struct task_struct *t,
 		       __func__, t->comm, t->pid, ret);
 		return ret;
 	}
+#ifdef CONFIG_EULEROS_LIVEPATCH_FORCE_PATCH
+	ret = klp_check_stack(&trace, 0, fn, data, enable, unloadable);
+#else
 	ret = klp_check_stack(&trace, 0, fn, data);
+#endif
 #endif
 	if (ret) {
 		pr_err("%s:%d check stack failed, ret=%d\n",
@@ -176,7 +206,11 @@ static int check_task_calltrace(struct task_struct *t,
 	return 0;
 }
 
+#ifdef CONFIG_EULEROS_LIVEPATCH_FORCE_PATCH
+static int do_check_calltrace(bool (*fn)(void *, int *, unsigned long, int, int *), void *data, int enable, int *unloadable)
+#else
 static int do_check_calltrace(bool (*fn)(void *, int *, unsigned long), void *data)
+#endif
 {
 	int ret = 0;
 	struct task_struct *g, *t;
@@ -185,20 +219,31 @@ static int do_check_calltrace(bool (*fn)(void *, int *, unsigned long), void *da
 	for_each_process_thread(g, t) {
 		if (klp_is_migration_thread(t->comm))
 			continue;
-
+#ifdef CONFIG_EULEROS_LIVEPATCH_FORCE_PATCH
+		ret = check_task_calltrace(t, fn, data, enable, unloadable);
+#else
 		ret = check_task_calltrace(t, fn, data);
+#endif
 		if (ret)
 			return ret;
 	}
 	for_each_online_cpu(cpu) {
+#ifdef CONFIG_EULEROS_LIVEPATCH_FORCE_PATCH
+		ret = check_task_calltrace(idle_task(cpu), fn, data, enable, unloadable);
+#else
 		ret = check_task_calltrace(idle_task(cpu), fn, data);
+#endif
 		if (ret)
 			return ret;
 	}
 	return 0;
 }
 
+#ifdef CONFIG_EULEROS_LIVEPATCH_FORCE_PATCH
+static bool check_module_calltrace(void *data, int *ret, unsigned long pc, int enable, int *unloadable)
+#else
 static bool check_module_calltrace(void *data, int *ret, unsigned long pc)
+#endif
 {
 	struct module *mod = (struct module *)data;
 
@@ -211,24 +256,46 @@ static bool check_module_calltrace(void *data, int *ret, unsigned long pc)
 }
 
 #ifdef CONFIG_LIVEPATCH_BREAKPOINT_NO_STOP_MACHINE
+#ifdef CONFIG_EULEROS_LIVEPATCH_FORCE_PATCH
+int arch_klp_check_task_calltrace(struct task_struct *t,
+				  bool (*fn)(void *, int *, unsigned long, int, int *),
+				  void *data)
+#else
 int arch_klp_check_task_calltrace(struct task_struct *t,
 				  bool (*fn)(void *, int *, unsigned long),
 				  void *data)
+#endif
 {
 	if (t == NULL)
 		return -EINVAL;
+#ifdef CONFIG_EULEROS_LIVEPATCH_FORCE_PATCH
+	return check_task_calltrace(t, fn, data, 0, NULL);
+#else
 	return check_task_calltrace(t, fn, data);
+#endif
 }
 #endif
 
+#ifdef CONFIG_EULEROS_LIVEPATCH_FORCE_PATCH
+int arch_klp_check_calltrace(bool (*check_func)(void *, int *, unsigned long, int, int *), void *data, int enable, int *unloadable)
+#else
 int arch_klp_check_calltrace(bool (*check_func)(void *, int *, unsigned long), void *data)
+#endif
 {
+#ifdef CONFIG_EULEROS_LIVEPATCH_FORCE_PATCH
+	return do_check_calltrace(check_func, data, enable, unloadable);
+#else
 	return do_check_calltrace(check_func, data);
+#endif
 }
 
 int arch_klp_module_check_calltrace(void *data)
 {
+#ifdef CONFIG_EULEROS_LIVEPATCH_FORCE_PATCH
+	return do_check_calltrace(check_module_calltrace, data, 0, NULL);
+#else
 	return do_check_calltrace(check_module_calltrace, data);
+#endif
 }
 
 int arch_klp_check_breakpoint(struct arch_klp_data *arch_data, void *old_func)
diff --git a/include/linux/livepatch.h b/include/linux/livepatch.h
index a5ef153ba..0d47ecb45 100644
--- a/include/linux/livepatch.h
+++ b/include/linux/livepatch.h
@@ -18,6 +18,14 @@
 
 #include <asm/livepatch.h>
 
+#ifdef CONFIG_EULEROS_LIVEPATCH_FORCE_PATCH
+enum klp_patch_state {
+   KLP_DISABLED,
+   KLP_ENABLED,
+   KLP_FORCEDISABLED,
+};
+#endif
+
 /* task patch states */
 #define KLP_UNDEFINED	-1
 #define KLP_UNPATCHED	 0
@@ -192,7 +200,11 @@ struct klp_patch {
 	struct list_head list;
 	struct kobject kobj;
 	struct list_head obj_list;
+#ifdef CONFIG_EULEROS_LIVEPATCH_FORCE_PATCH
+	enum klp_patch_state state;
+#else
 	bool enabled;
+#endif
 	bool forced;
 	struct work_struct free_work;
 	struct completion finish;
@@ -261,9 +273,15 @@ typedef int (*klp_add_func_t)(struct list_head *func_list,
 			       const char *func_name, int force);
 
 struct walk_stackframe_args {
-	void *data;
 	int ret;
+	void *data;
+#ifdef CONFIG_EULEROS_LIVEPATCH_FORCE_PATCH
+	bool (*check_func)(void *data, int *ret, unsigned long pc, int enable, int *unloadable);
+	int enable;
+	int unloadable;
+#else
 	bool (*check_func)(void *data, int *ret, unsigned long pc);
+#endif
 };
 
 #endif
diff --git a/kernel/livepatch/Kconfig b/kernel/livepatch/Kconfig
index a59cbb650..03397ccb7 100644
--- a/kernel/livepatch/Kconfig
+++ b/kernel/livepatch/Kconfig
@@ -109,5 +109,17 @@ config LIVEPATCH_BREAKPOINT_NO_STOP_MACHINE
 	  service downtime.
 	  Say N if you are unsure.
 
+config EULEROS_LIVEPATCH_FORCE_PATCH
+	bool "support force enable or disable patch"
+	depends on LIVEPATCH && LIVEPATCH_STOP_MACHINE_CONSISTENCY
+	default y
+	help
+	  some functions in IO path are called frequently, in this case,
+	  patches can not be enabled due to stack check failing. We use
+	  the klp_func->force to indicate functions that can be
+	  enabled/disabled without stack check. Also, we add a new patch
+	  state forcedisabled to indicate the patch being disabled and
+	  could not being rmmod.
+
 endmenu
 endif
diff --git a/kernel/livepatch/core.c b/kernel/livepatch/core.c
index 23cb14a8d..ace3cf3c8 100644
--- a/kernel/livepatch/core.c
+++ b/kernel/livepatch/core.c
@@ -475,9 +475,15 @@ static ssize_t enabled_store(struct kobject *kobj, struct kobj_attribute *attr,
 {
 	struct klp_patch *patch;
 	int ret;
+#ifdef CONFIG_EULEROS_LIVEPATCH_FORCE_PATCH
+	unsigned long enabled;
+
+	ret = kstrtoul(buf, 10, &enabled);
+#else
 	bool enabled;
 
 	ret = kstrtobool(buf, &enabled);
+#endif
 	if (ret)
 		return ret;
 
@@ -494,7 +500,11 @@ static ssize_t enabled_store(struct kobject *kobj, struct kobj_attribute *attr,
 		goto out;
 	}
 
+#ifdef CONFIG_EULEROS_LIVEPATCH_FORCE_PATCH
+	if (patch->state == enabled) {
+#else
 	if (patch->enabled == enabled) {
+#endif
 		/* already in requested state */
 		ret = -EINVAL;
 		goto out;
@@ -520,7 +530,11 @@ static ssize_t enabled_show(struct kobject *kobj,
 	struct klp_patch *patch;
 
 	patch = container_of(kobj, struct klp_patch, kobj);
+#ifdef CONFIG_EULEROS_LIVEPATCH_FORCE_PATCH
+	return snprintf(buf, PAGE_SIZE-1, "%d\n", patch->state);
+#else
 	return snprintf(buf, PAGE_SIZE-1, "%d\n", patch->enabled);
+#endif
 }
 
 #ifdef CONFIG_LIVEPATCH_PER_TASK_CONSISTENCY
@@ -590,10 +604,19 @@ static int state_show(struct seq_file *m, void *v)
 	seq_puts(m, "-----------------------------------------------\n");
 	mutex_lock(&klp_mutex);
 	list_for_each_entry(patch, &klp_patches, list) {
+#ifdef CONFIG_EULEROS_LIVEPATCH_FORCE_PATCH
+		if (patch->state == KLP_ENABLED)
+			state = "enabled";
+		else if (patch->state == KLP_DISABLED)
+			state = "disabled";
+		else
+			state = "forcedisabled";
+#else
 		if (patch->enabled)
 			state = "enabled";
 		else
 			state = "disabled";
+#endif
 
 		seq_printf(m, "%-5d\t%-26s\t%-8s\n", ++index,
 				patch->mod->name, state);
@@ -1177,7 +1200,11 @@ static void klp_init_patch_early(struct klp_patch *patch)
 	INIT_LIST_HEAD(&patch->list);
 	INIT_LIST_HEAD(&patch->obj_list);
 	kobject_init(&patch->kobj, &klp_ktype_patch);
+#ifdef CONFIG_EULEROS_LIVEPATCH_FORCE_PATCH
+	patch->state = KLP_DISABLED;
+#else
 	patch->enabled = false;
+#endif
 	patch->forced = false;
 	INIT_WORK(&patch->free_work, klp_free_patch_work_fn);
 	init_completion(&patch->finish);
@@ -1329,7 +1356,11 @@ static int __klp_disable_patch(struct klp_patch *patch)
 	return 0;
 }
 #elif defined(CONFIG_LIVEPATCH_STOP_MACHINE_CONSISTENCY)
+#ifdef CONFIG_EULEROS_LIVEPATCH_FORCE_PATCH
+int __weak arch_klp_check_calltrace(bool (*fn)(void *, int *, unsigned long, int, int *), void *data, int enable, int *unloadable)
+#else
 int __weak arch_klp_check_calltrace(bool (*fn)(void *, int *, unsigned long), void *data)
+#endif
 {
 	return -EINVAL;
 }
@@ -1388,6 +1419,18 @@ int __weak arch_klp_check_activeness_func(struct klp_func *func, int enable,
 				return ret;
 		}
 	} else {
+#ifdef CONFIG_EULEROS_LIVEPATCH_FORCE_PATCH
+		if (func->force) {
+			func_addr = (unsigned long)func->new_func;
+			func_size = func->new_size;
+			ret = add_func(func_list, func_addr,
+					func_size, func->old_name, func->force);
+			if (ret)
+				return ret;
+			return 0;
+		}
+#endif
+
 #ifdef CONFIG_PREEMPTION
 		/*
 		 * No scheduling point in the replacement instructions. Therefore,
@@ -1440,7 +1483,11 @@ struct actv_func {
 	int force;
 };
 
+#ifdef CONFIG_EULEROS_LIVEPATCH_FORCE_PATCH
+static bool check_func_list(void *data, int *ret, unsigned long pc, int enable, int *unloadable)
+#else
 static bool check_func_list(void *data, int *ret, unsigned long pc)
+#endif
 {
 	struct list_head *func_list = (struct list_head *)data;
 	struct actv_func *func = NULL;
@@ -1448,8 +1495,33 @@ static bool check_func_list(void *data, int *ret, unsigned long pc)
 	list_for_each_entry(func, func_list, list) {
 		*ret = klp_compare_address(pc, func->func_addr, func->func_name,
 				klp_size_to_check(func->func_size, func->force));
+#ifdef CONFIG_EULEROS_LIVEPATCH_FORCE_PATCH
+		if (*ret) {
+			if (!func->force) {
+#ifdef CONFIG_X86
+				*ret = -EAGAIN;
+#endif
+				return false;
+			}
+#ifdef CONFIG_X86
+			if (enable) {
+				return true;
+			} else {
+				*ret = -EBUSY;
+				return false;
+			}
+#else
+			if (!enable) {
+				*unloadable = 0;
+			}
+			*ret = 0;
+			return true;
+#endif
+		}
+#else
 		if (*ret)
 			return false;
+#endif
 	}
 	return true;
 }
@@ -1500,7 +1572,11 @@ static int klp_check_activeness_func(struct klp_patch *patch, int enable,
 	return 0;
 }
 
+#ifdef CONFIG_EULEROS_LIVEPATCH_FORCE_PATCH
+static int klp_check_calltrace(struct klp_patch *patch, int enable, int *unloadable)
+#else
 static int klp_check_calltrace(struct klp_patch *patch, int enable)
+#endif
 {
 	int ret = 0;
 	LIST_HEAD(func_list);
@@ -1514,7 +1590,11 @@ static int klp_check_calltrace(struct klp_patch *patch, int enable)
 	if (list_empty(&func_list))
 		goto out;
 
+#ifdef CONFIG_EULEROS_LIVEPATCH_FORCE_PATCH
+	ret = arch_klp_check_calltrace(check_func_list, (void *)&func_list, enable, unloadable);
+#else
 	ret = arch_klp_check_calltrace(check_func_list, (void *)&func_list);
+#endif
 
 out:
 	free_func_list(&func_list);
@@ -1584,13 +1664,30 @@ unlock:
 /*
  * This function is called from stop_machine() context.
  */
+#ifdef CONFIG_EULEROS_LIVEPATCH_FORCE_PATCH
+static int disable_patch(struct klp_patch *patch, int unloadable)
+{
+	if (patch->state == KLP_FORCEDISABLED)
+		goto modput;
+#else
 static int disable_patch(struct klp_patch *patch)
 {
+#endif
 	pr_notice("disabling patch '%s'\n", patch->mod->name);
 
 	klp_unpatch_objects(patch);
+#ifndef CONFIG_EULEROS_LIVEPATCH_FORCE_PATCH
 	patch->enabled = false;
 	module_put(patch->mod);
+#else
+modput:
+	if (unloadable) {
+		patch->state = KLP_DISABLED;
+		module_put(patch->mod);
+	} else
+		patch->state = KLP_FORCEDISABLED;
+#endif
+
 	return 0;
 }
 
@@ -1598,6 +1695,9 @@ int klp_try_disable_patch(void *data)
 {
 	int ret = 0;
 	struct patch_data *pd = (struct patch_data *)data;
+#ifdef CONFIG_EULEROS_LIVEPATCH_FORCE_PATCH
+	int unloadable = 1;
+#endif
 
 	if (atomic_inc_return(&pd->cpu_count) == 1) {
 		struct klp_patch *patch = pd->patch;
@@ -1607,12 +1707,20 @@ int klp_try_disable_patch(void *data)
 			return -EINVAL;
 		}
 
+#ifdef CONFIG_EULEROS_LIVEPATCH_FORCE_PATCH
+		ret = klp_check_calltrace(patch, 0, &unloadable);
+#else
 		ret = klp_check_calltrace(patch, 0);
+#endif
 		if (ret) {
 			atomic_inc(&pd->cpu_count);
 			return ret;
 		}
+#ifdef CONFIG_EULEROS_LIVEPATCH_FORCE_PATCH
+		ret = disable_patch(patch, unloadable);
+#else
 		ret = disable_patch(patch);
+#endif
 		if (ret) {
 			atomic_inc(&pd->cpu_count);
 			return ret;
@@ -1885,13 +1993,21 @@ static int __klp_disable_patch(struct klp_patch *patch)
 		.cpu_count = ATOMIC_INIT(0),
 	};
 
+#ifdef CONFIG_EULEROS_LIVEPATCH_FORCE_PATCH
+	if (WARN_ON(patch->state == KLP_DISABLED))
+#else
 	if (WARN_ON(!patch->enabled))
+#endif
 		return -EINVAL;
 
 #ifdef CONFIG_LIVEPATCH_STACK
 	/* enforce stacking: only the last enabled patch can be disabled */
 	if (!list_is_last(&patch->list, &klp_patches) &&
+#ifdef CONFIG_EULEROS_LIVEPATCH_FORCE_PATCH
+	    list_next_entry(patch, list)->state == KLP_ENABLED) {
+#else
 	    list_next_entry(patch, list)->enabled) {
+#endif
 		pr_err("only the last enabled patch can be disabled\n");
 		return -EBUSY;
 	}
@@ -2053,11 +2169,23 @@ static int enable_patch(struct klp_patch *patch, bool rollback)
 	pr_notice_once("tainting kernel with TAINT_LIVEPATCH\n");
 	add_taint(TAINT_LIVEPATCH, LOCKDEP_STILL_OK);
 
+#ifdef CONFIG_EULEROS_LIVEPATCH_FORCE_PATCH
+	if (KLP_ENABLED != patch->state) {
+#else
 	if (!patch->enabled) {
+#endif
+#ifdef CONFIG_EULEROS_LIVEPATCH_FORCE_PATCH
+	if (patch->state == KLP_DISABLED && !try_module_get(patch->mod))
+#else
 		if (!try_module_get(patch->mod))
+#endif
 			return -ENODEV;
 
+#ifdef CONFIG_EULEROS_LIVEPATCH_FORCE_PATCH
+		patch->state = KLP_ENABLED;
+#else
 		patch->enabled = true;
+#endif
 
 		pr_notice("enabling patch '%s'\n", patch->mod->name);
 	}
@@ -2077,7 +2205,11 @@ static int enable_patch(struct klp_patch *patch, bool rollback)
 	return 0;
 
 disable:
+#ifdef CONFIG_EULEROS_LIVEPATCH_FORCE_PATCH
+	disable_patch(patch, 1);
+#else
 	disable_patch(patch);
+#endif
 	return ret;
 }
 
@@ -2094,7 +2226,11 @@ int klp_try_enable_patch(void *data)
 			return -EINVAL;
 		}
 
+#ifdef CONFIG_EULEROS_LIVEPATCH_FORCE_PATCH
+		ret = klp_check_calltrace(patch, 1, NULL);
+#else
 		ret = klp_check_calltrace(patch, 1);
+#endif
 		if (ret) {
 			atomic_inc(&pd->cpu_count);
 			return ret;
@@ -2146,9 +2282,15 @@ static bool klp_use_breakpoint(struct klp_patch *patch)
 #include <linux/sched/task.h>
 #include "../sched/sched.h"
 
+#ifdef CONFIG_EULEROS_LIVEPATCH_FORCE_PATCH
+int __weak arch_klp_check_task_calltrace(struct task_struct *t,
+					 bool (*fn)(void *, int *, unsigned long, int, int *),
+					 void *data)
+#else
 int __weak arch_klp_check_task_calltrace(struct task_struct *t,
 					 bool (*fn)(void *, int *, unsigned long),
 					 void *data)
+#endif
 {
 	return -EINVAL;
 }
@@ -2382,13 +2524,21 @@ static int __klp_enable_patch(struct klp_patch *patch)
 		.rollback = true,
 	};
 
+#ifdef CONFIG_EULEROS_LIVEPATCH_FORCE_PATCH
+	if (WARN_ON(patch->state == KLP_ENABLED))
+#else
 	if (WARN_ON(patch->enabled))
+#endif
 		return -EINVAL;
 
 #ifdef CONFIG_LIVEPATCH_STACK
 	/* enforce stacking: only the first disabled patch can be enabled */
 	if (patch->list.prev != &klp_patches &&
+#ifdef CONFIG_EULEROS_LIVEPATCH_FORCE_PATCH
+	    list_prev_entry(patch, list)->state == KLP_DISABLED) {
+#else
 	    !list_prev_entry(patch, list)->enabled) {
+#endif
 		pr_err("only the first disabled patch can be enabled\n");
 		return -EBUSY;
 	}
@@ -2521,7 +2671,11 @@ int klp_unregister_patch(struct klp_patch *patch)
 		goto out;
 	}
 
+#ifdef CONFIG_EULEROS_LIVEPATCH_FORCE_PATCH
+	if (patch->state == KLP_ENABLED) {
+#else
 	if (patch->enabled) {
+#endif
 		ret = -EBUSY;
 		goto out;
 	}
@@ -2598,7 +2752,11 @@ void klp_unpatch_replaced_patches(struct klp_patch *new_patch)
 		if (old_patch == new_patch)
 			return;
 
+#ifdef CONFIG_EULEROS_LIVEPATCH_FORCE_PATCH
+		old_patch->state = KLP_DISABLED;
+#else
 		old_patch->enabled = false;
+#endif
 		klp_unpatch_objects(old_patch);
 	}
 }
-- 
2.27.0

